api/src/getStartDate.ts

```diff
@@ -0,0 +1,144 @@
+        // Getting the difference in time so Freerooms gets the most recent term start date when
+        // it starts.
+        const timeDelta = currDate.valueOf() - termStartDateObj.valueOf();
+
+        if (timeDelta >= 0) {
+            // If the data is more than a day old, we will return the term start date that is for the
+            // new term, else we get the date for the old term.
+            res.send(termStartDate);
+        } else {
+            let dummyTerm = termArray.findIndex((term) => term == currTerm) - 1;
+            // Checking if the term is the first term, we can have case when index out of bounds.
+            if (dummyTerm < 0) {
+                dummyTerm = 0; // Sanity check to make sure the array does not break
+                // Only time this will happen is when the next year's summer starts over.
+            }
+            const oldTermStartDate =
+                data.timetableData[termArray[dummyTerm]][FIRST_COURSE]["classes"][0]["termDates"][
+                    "start"
+                ];
+            res.send(oldTermStartDate);
+        }
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the latest term start date from data.json
+ */
+const getStartDateNotangles = (req: express.Request, res: express.Response) => {
+    try {
+        // Get the latest term and check the first class of the term
+        // The assumption is that the starting date is the start date of the first class
+        // of the term.
+        const termStartDate =
+            data.timetableData[term()][FIRST_COURSE]["classes"][0]["termDates"]["start"];
+        res.send(termStartDate);
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the latest term start date from data.json
+ */
+const getAvailableTermData = (req: express.Request, res: express.Response) => {
+    try {
+        res.send(term());
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the term which has data from data.json, this term
+ * is the latest term with data. This is exclusively for Notangles, so
+ * Freerooms please do not use it!
+ * @returns the latest term number
+ */
+const term = () => {
+    try {
+        let term: string = "";
+        for (let termId of termArray) {
+            let timetableData = data.timetableData;
+            // Check if the termId exists in the timetableData and the fields are actually
+            // present. If they are, return the term as it is.
+            if (timetableData.hasOwnProperty(termId)) {
+                if (!isClassFound(termId, "ECON1101")) {
+                    break;
+                }
+                term = termId;
+            }
+        }
+
+        return term;
+    } catch (e) {
+        console.log("There was an error getting the most updated term!");
+    }
+};
+
+const isClassFound = (termId: string, courseCode: string) => {
+    let timetableData = data.timetableData;
```

Comment:
This logic is too long-winded and can be replaced by one of the array functions.